### DISCUSSION TITLE
Writing to file

### DISCUSSION THREAD

**Discussion by Pandagineer** [bitcraze/discussions#338] [Q&A] [answered]
Hello. I would like to write the contents of a variable to a file (rather than to the Console). The code is below. As you can see, at this point I'm not utilizing the Crazyflie... that will be the next step. When I make, it gives me an error (make log is below the code).

Thanks!

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include "app.h"
#include "FreeRTOS.h"
#include "task.h"
#include "deck_constants.h"
#include "deck_analog.h"
#include "debug.h"

void appMain() 
  {int num = 14;
  FILE *fptr;
  fptr = fopen("/home/bitcraze/Desktop/Personal_Projects/program.txt","w");
  fprintf(fptr,"%d",num);
  fclose(fptr);
}

```

bitcraze@ubuntu:~/Desktop/Personal_Projects/Measure_Voltage$ make
make KBUILD_OUTPUT=/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build -C /home/bitcraze/Desktop/projects/crazyflie-firmware OOT=/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage EXTRA_CFLAGS="" KCONFIG_ALLCONFIG=/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/app-config alldefconfig
make[1]: Entering directory '/home/bitcraze/projects/crazyflie-firmware'
make[2]: Entering directory '/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build'
  GEN     ./Makefile
scripts/kconfig/conf  --alldefconfig Kconfig

 configuration written to .config

make[2]: Leaving directory '/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build'
make[1]: Leaving directory '/home/bitcraze/projects/crazyflie-firmware'
make KBUILD_OUTPUT=/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build -C /home/bitcraze/Desktop/projects/crazyflie-firmware OOT=/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage EXTRA_CFLAGS="" -j 12
make[1]: Entering directory '/home/bitcraze/projects/crazyflie-firmware'
make[2]: Entering directory '/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build'
  GEN     ./Makefile
scripts/kconfig/conf  --silentoldconfig Kconfig
  Using /home/bitcraze/projects/crazyflie-firmware as source for kernel
  GEN     ./Makefile
  CHK     include/generated/version.h
  CC      /home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/src/Measure_Voltage.o
  LD      /home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/src/built-in.o
  LD      /home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/built-in.o
  LD      firmware
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-fstatr.o): in function `_fstat_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/fstatr.c:55: undefined reference to `_fstat'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-openr.o): in function `_open_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/openr.c:50: undefined reference to `_open'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-sbrkr.o): in function `_sbrk_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/sbrkr.c:51: undefined reference to `_sbrk'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-writer.o): in function `_write_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/writer.c:49: undefined reference to `_write'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-closer.o): in function `_close_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/closer.c:47: undefined reference to `_close'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-isattyr.o): in function `_isatty_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/isattyr.c:52: undefined reference to `_isatty'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-lseekr.o): in function `_lseek_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/lseekr.c:49: undefined reference to `_lseek'
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc.a(lib_a-readr.o): in function `_read_r':
/build/newlib-CVVEyx/newlib-3.3.0/build/arm-none-eabi/thumb/v7e-m+fp/hard/newlib/libc/reent/../../../../../../../../newlib/libc/reent/readr.c:49: undefined reference to `_read'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/kbuild/Makefile.kbuild:572: firmware] Error 1
make[2]: Leaving directory '/home/bitcraze/Desktop/Personal_Projects/Measure_Voltage/build'
make[1]: *** [tools/kbuild/Makefile.kbuild:150: sub-make] Error 2
make[1]: Leaving directory '/home/bitcraze/projects/crazyflie-firmware'
make: *** [/home/bitcraze/Desktop/projects/crazyflie-firmware/tools/make/oot.mk:23: all] Error 2

**Comment by knmcguire** [comment, marked as answer]
Hi!

Functions like fopen or fprintf are not possible on a crazyflie. Also it does not use a file system like a regular computer so what you would like to do will not  be possible on an embedded system like the crazyflie.

you will only be able to write a file with the [uSD deck ](https://www.bitcraze.io/products/micro-sd-card-deck/) driver. Please check the usage tab of the product page.

**Reply by Pandagineer** [reply]
This has been resolved. My goal was to collect data from the analog pins, along with a timestamp, and then write those data to a .csv file. Later I watched your YouTube video on how to add a variable to Log TOC, plot it, and save it as .csv (which includes the relative time). I successfully did this. Thank you!

**Reply by knmcguire** [reply]
that is indeed how it should be done with the crazyflie framework! good you figured it out.
